### PR TITLE
[TensorPipe] Register multiplexing channel over UV

### PR DESCRIPTION
**Pull request by lw**
Stack from [ghstack](https://github.com/ezyang/ghstack):
* **#40389 [TensorPipe] Register multiplexing channel over UV**

The `mpt_uv` channel MultiPlexes over a Transport, namely the UV one. What this means is that it takes a tensor, chunks it into equal parts and sends each of them on a separate UV connection, each running in a separate UV loop. Thus they each have their own socket and thread. This allows them to reach bandwidths that go beyond what a simple single-threaded approach can do, which is necessary to reach the high bandwidths of some modern NICs.

Differential Revision: [D22144380](https://our.internmc.facebook.com/intern/diff/D22144380/)

**NOTE FOR REVIEWERS**: This PR has internal Facebook specific changes or comments, please review them on [Phabricator](https://our.internmc.facebook.com/intern/diff/D22144380/)!

### CODE REVIEW

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp [hunk 1]
```diff
@@ -128,6 +129,34 @@ C10_REGISTER_CREATOR(TensorPipeChannelRegistry, cma, makeCmaChannel);
 
 #endif
 
+constexpr static int kNumUvThreads = 16;
```
Comment (mrshenli) [inline torch/csrc/distributed/rpc/tensorpipe_agent.cpp:132]: This means we will also need expose channel configs in the future?

**Review by mrshenli — APPROVED** [review]
This LGTM. Shall we add some tests?

**Comment by lw** [comment]
Yes, some better testing of the various transports/channels is needed. You may remember we recently discussed some further refactoring of the RPC tests, I think it can be done at the same time. I want to give that a shot, after the v1.6 branch cut.

File: torch/csrc/distributed/rpc/tensorpipe_agent.cpp (hunk 1 above)
Comment (lw) [reply torch/csrc/distributed/rpc/tensorpipe_agent.cpp:132]: Indeed, we should have a way to make this configurable. The default I'm putting here, 16, should be enough to cover NICs up to 100 Gbps. On the other hand, it means that whoever is using TensorPipe ends up spawning 16 threads for this channel, even if their NIC is smaller than that. Ideally there should be an auto-sizing mechanism in TensorPipe, or at least a system to turn these on only when needed.

For v1.6, as multiplexing is one of the main benefits of TensorPipe over Gloo, I'm inclined to leave it always on with a large enough number of threads, to ensure everyone can benefit from this. But we definitely need to revisit it.

**Pull request closed without merging**